code/controllers/autofire.dm

```diff
@@ -0,0 +1,225 @@
+		return
+
+	// Otherwise it's a simple insertion into the double-linked list
+	if (bucket_head.next)
+		next = bucket_head.next
+		next.prev = src
+	
+	bucket_head.next = src
+	prev = bucket_head
+	
+	//Something went wrong, probably a lag spike or something. To prevent infinite loops, we reschedule it to a another next fire
+	if(prev == src)
+		next_fire = next_fire += 1
+		schedule_shot()
+
+///Handle the firing of the autofire component
+/datum/component/automatedfire/proc/process_shot()
+	return
+
+#undef BUCKET_LEN
+#undef BUCKET_POS
+
+/obj/structure/turret_debug
+	name = "debug turret"
+	///What kind of ammo it uses
+	var/datum/ammo/ammo
+	///Its target
+	var/atom/target 
+	///At wich rate it fires in ticks
+	var/firerate = 5.5
+
+/obj/structure/turret_debug/fast
+	name = "debug turret fast"
+	firerate = 1
+
+/obj/structure/turret_debug/super_fast
+	name = "debug turret super fast"
+	firerate = 0.5
+
+/obj/structure/turret_debug/slow
+	name = "debug turret slow"
+	firerate = 25
+
+/obj/structure/turret_debug/Initialize()
+	. = ..()
+	ammo = GLOB.ammo_list[/datum/ammo/xeno/acid]
+	target = locate(x+5, y, z)
+	AddComponent(/datum/component/automatedfire/automatic_shoot_at, firerate)
+	RegisterSignal(src, COMSIG_AUTOMATIC_SHOOTER_SHOT_FIRED, .proc/shoot)
+	SEND_SIGNAL(src, COMSIG_AUTOMATIC_SHOOTER_START_SHOOTING_AT)
+	var/static/number = 1
+	name = "[name] [number]"
+	number++
+
+///Create the projectile
+/obj/structure/turret_debug/proc/shoot()
+	SIGNAL_HANDLER
+	var/obj/projectile/newshot = new(loc)
+	newshot.generate_bullet(ammo)
+	newshot.permutated += src
+	newshot.fire_at(target, src, null, ammo.max_range, ammo.shell_speed)
+
+/datum/component/automatedfire/automatic_shoot_at
+	///The delay between each shot in ticks
+	var/shot_delay = 5
+	///If we are shooting
+	var/shooting = FALSE
+	///Last time it fired
+	var/last_fire
+
+/datum/component/automatedfire/automatic_shoot_at/Initialize(_shot_delay)
+	. = ..()
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+	shooter = parent
+	shot_delay = _shot_delay
+	RegisterSignal(parent, COMSIG_AUTOMATIC_SHOOTER_START_SHOOTING_AT, .proc/start_shooting)
+	RegisterSignal(parent, COMSIG_AUTOMATIC_SHOOTER_STOP_SHOOTING_AT, .proc/stop_shooting)
+
+///Signal handler for starting the autoshooting at something
+/datum/component/automatedfire/automatic_shoot_at/proc/start_shooting()
+	SIGNAL_HANDLER
+	next_fire = world.time
+	if(!shooting)
+		shooting = TRUE
+		schedule_shot()
+
+
+///Signal handler for stoping the shooting
+/datum/component/automatedfire/automatic_shoot_at/proc/stop_shooting(datum/source)
+	SIGNAL_HANDLER
+	shooting = FALSE
+
+/datum/component/automatedfire/automatic_shoot_at/process_shot()
+	if(!shooting)
+		return
+	SEND_SIGNAL(shooter, COMSIG_AUTOMATIC_SHOOTER_SHOT_FIRED)
+	next_fire = world.time + shot_delay
+	if(world.time - last_fire != shot_delay)
+		message_admins("the delay between shots was [world.time - last_fire] ticks instead of [shot_delay]")
```

Comment:
Debug stuff comitted

Reply:
It's for the debug turrets only, so it's no big deal